tiny functions added/implemented ( #139 )

lvntky opened the pull request:
Hi, I added 2 tiny functions due the issue #139 . I did not implement the `generate_symmetric_key ` function because I'm not sure that I'm on the right track, and I don't want to affect the project badly. And sorry if I do something silly :) this is my first ever PR. Anyway, if the implementation goes right I can start the implementing `generate_symmetric_key` function similar to  [this](https://github.com/mc2-project/mc2/blob/master/python-package/mc2client/core.py#L521) function.

chester-leung commented:
Hi @lvntky this looks great! There's actually no need to implement a `generate_symmetric_key()` function in C++. The C++ should not be user facing - the user should only be calling Python functions.

We can continue to use [this function](https://github.com/mc2-project/mc2/blob/master/python-package/mc2client/core.py#L585), but replace the `num_bytes` variable with the result of calling `_LIB.cipher_key_size()` in Python (this calls the C++ function you wrote from Python). This is the only change necessary to fix this issue.

We should then remove the parameter from the `generate_symmetric_key()` function since we no longer need it to be passed in. Let me know if this makes sense or if you have any other questions!

lvntky commented:
hi @chester-leung thank you for this great and informative explanation. So as I understand I make only changes on `generate_symmetric_key()` on the core.py file right? If it is right I will start working on it ASAP. Also thank you very much for your friendly and helpful response friend  :pray:

chester-leung commented:
@lvntky yes that's correct. You should only have to make a change in the `generate_symmetric_key()` function in `core.py` to call the C++ `cipher_key_size()` function you added in C++. Thank you!

lvntky commented:
@chester-leung thank you very much! I'm on to it then

lvntky commented:
hi @chester-leung I made a new contribution and small changes. I deleted the `extern "C" int generate_symmetric_key(cipher_key_size()) {}` in the c_api.cpp file since we don't need, and I change `num_bytes` variable with `_LIB.cipher_key_size()` in the core.py file. Sorry if I made something silly, as I said this is my first ever PR, and I really want to contribute to this project. And also thank you for your patience, informative and friendly instructions. If I made something wrong let me know and change it. Best wishes :pray:

chester-leung commented:
@lvntky this looks great! Left a comment.

Please also fix the build (I think you're missing a semicolon [here](https://github.com/mc2-project/mc2/pull/146/files#diff-6a27bb364a4de78af158135b3d8fc225dd432d2c8d15c06270b42f38e6e6f319R146)) and also run our pre-commit hooks discussed in step 4 [here](https://github.com/mc2-project/mc2/blob/master/CONTRIBUTING.md#to-contribute-a-patch) to format your code.

chester-leung commented on python-package/mc2client/core.py:
```diff
@@ -603,7 +604,7 @@ def generate_symmetric_key(num_bytes=32):
         return
 
-    key = AESGCM.generate_key(bit_length=num_bytes * 8)
+    key = AESGCM.generate_key(bit_length= _LIB.cipher_key_size() * 8)
```
Can you please create a variable `num_bytes` and set it to `_LIB.cipher_key_size()` for clarity? This will add one more line but should make things easier to read.

lvntky replied:
of course I'm doing it right now

lvntky commented:
hi @chester-leung I add `num_byte` variable end assign it to `_LIB.cipher_key_size()` for better readability, as you said so.

chester-leung commented:
Thanks! Please update us once you've fixed the build and reformatted.

lvntky commented:
hi @chester-leung  thank you for your patience and friendly instructions. i added a semicolon to cpp function and I run the` pre-commit`. Also, I added some comments to functions, I can delete it if you want. But I'm a little confused about the build, should I build with kubernetes?

chester-leung commented on python-package/mc2client/core.py:
```diff
@@ -591,6 +591,16 @@ def generate_symmetric_key(num_bytes=32):
     """
+    
+    """
+    Due the issue #139, 
```
Please remove these comments

chester-leung commented on src/c_api.cpp:
```diff
@@ -142,3 +142,7 @@ extern "C" void opaque_decrypt_data(char **encrypted_files,
     *result = status;
 }
+
+// function created for to not to allow user specifies cipher key
```
please remove this comment

chester-leung commented:
Pre-commit essentially installs some git "hooks" that will reformat your code before you commit. I'm not sure if the code has been formatted correctly - in particular, there likely shouldn't be a space between `=` and `num_bytes` [here](https://github.com/mc2-project/mc2/pull/146/files#diff-7b1d6e7149d7ff0346c5d9e77a0dc1d480c15b512a274979a1932f58f8f0b52fR616)

Here are some instructions
```
# Install pre-commit
pip3 install pre-commit

# in the root mc2 directory, install the pre-commit hooks
pre-commit install
```

Then, after you make a change to the code and `git add/commit`, you should see the following:
![image](https://user-images.githubusercontent.com/18539530/121945341-86cdd180-cd08-11eb-85bc-bf0394036f11.png)


By build, I meant these checks
![image](https://user-images.githubusercontent.com/18539530/121944376-7701bd80-cd07-11eb-9d58-36db98b27957.png)

It looks like you're passing them now, great! Eventually we'll want to also add a linter check to enforce formatting (and to ensure that all contributors have run the pre-commit formatting hooks.

lvntky commented:
hi @chester-leung i deleted the comments and run pre-commit. here is the screenshot: 
![image](https://user-images.githubusercontent.com/42411502/121953009-ac3cf880-cd65-11eb-8656-72a11ed06547.png)

chester-leung commented:
This looks good! Just FYI, when you run pre-commit and it modifies some files, you'll have to readd the modified files to the git staging area. Formatting looks good for now though!

Once the checks pass, we can merge this in. Thanks for your contribution @lvntky !

lvntky commented:
thank you very much for your friendly and informative information @chester-leung i am really glad to my first contribution is for mc2. I will hunting the issues again :smile:

chester-leung commented:
Sounds great! @lvntky if you're interested, #147 and #148 would be good to do next. Thank you again!

lvntky commented:
yup I am totally interested friend, i have small work into job after that i dive to them 😄
